src/transformStateWithClones.js

```diff
@@ -7,7 +7,34 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const statesArray = [];
+
+  let currentState = { ...state };
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        action.keysToRemove.forEach((key) => {
+          delete currentState[key];
+        });
+        break;
+
+      default:
+        break;
+    }
+
+    statesArray.push(currentState);
```

Comment:
You are pushing the same `currentState` object reference into `statesArray` on each iteration. This will result in all elements of `statesArray` pointing to the same object, which is the final state after all actions. Instead, you should push a copy of `currentState` to ensure each element in the array is a distinct object representing the state at that point in time.